### PR TITLE
Added Local Authorities tab

### PR DESCRIPTION
**Pull request by Majid-justice**
https://tools.hmcts.net/jira/browse/FACT-557

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/main/assets/js/local-authorities.ts [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        this.scrollable();
+      }).fail(response =>
+        AjaxErrorHandler.handleError(response, 'POST local authorities failed.'));
+    });
+  }
+
+  private displayLocalAuthoritiesTab(): void{
+
+    const isSuperAdmin = Boolean($('#isSuperAdmin').val() === 'true');
+    if(!isSuperAdmin){
```
Comment (amstevenson) [inline src/main/assets/js/local-authorities.ts:94]: I think we use middleware to check if the user is a super admin or not, before the tab is displayed? 

```
  app.get('/courts/:slug/postcodes', isSuperAdmin, app.locals.container.cradle.postcodesController.get);
  app.post('/courts/:slug/postcodes', isSuperAdmin, app.locals.container.cradle.postcodesController.post);
```

 Guessing that maybe its the case that we put this here as there a subset of the page that requires super admin rights? If so then not a problem :-)

File: src/main/app/controller/courts/LocalAuthoritiesController.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  LocalAuthoritiesPageData,
+  LocalAuthority,
+  LocalAuthorityItem
+} from '../../../types/LocalAuthority';
+import {AreaOfLaw} from '../../../types/AreaOfLaw';
+
+import {SelectItem} from '../../../types/CourtPageData';
+import {CSRF} from '../../../modules/csrf';
+
+export enum familyAreaOfLaw {
```
Comment (amstevenson) [inline src/main/app/controller/courts/LocalAuthoritiesController.ts:16]: I think generally if an enum is public, it should be in its own file. If it's private then it's cool. Would you mind either making it private or putting it in an enum folder of sorts? Thanks. Happy to discuss

File: src/main/app/controller/courts/LocalAuthoritiesController.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+  adoption = 'Adoption',
+  children = 'Children',
+  civilPartnership = 'Civil partnership',
+  divorce = 'Divorce'
+}
+@autobind
+export class LocalAuthoritiesController {
+
+
```
Comment (amstevenson) [inline src/main/app/controller/courts/LocalAuthoritiesController.ts:25]: Think there are a few places in the code where there may be more than one space, might be worth making it so there is only one space

File: src/main/app/controller/courts/LocalAuthoritiesController.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    req: AuthedRequest,
+    res: Response,
+    updated = false,
+    error = '',
+    selectedAreaOfLaw = '',
+    courtLocalAuthorities: LocalAuthority[] = null): Promise<void> {
+    const slug: string = req.params.slug as string;
+    const areaOfLaw: string = req.params.areaOfLaw as string;
+
+    let areasOfLaw: AreaOfLaw[] = null;
```
Comment (amstevenson) [inline src/main/app/controller/courts/LocalAuthoritiesController.ts:80]: This confuses me a little. By declaring: 

`let areasOfLaw: AreaOfLaw[] = null;` - it's always null on declaration 

```
    if (!areasOfLaw ) { <- always null by default as declared above, so goes in here 

        set areas of law 

    }

    if (areasOfLaw ){
         areas of law is always set, unless there is an error
    }
```

Is it better to do something like: 

```
let areasOfLaw: AreaOfLaw[] = null;
await req.scope.cradle.api.getCourtAreasOfLaw(slug)
        .then((value: AreaOfLaw[]) => areasOfLaw = value)
        .catch(() => {

        error += this.getCourtAreasOfLawErrorMsg;
        areasOfLaw = this.checkFamilyAreasOfLaw(areasOfLaw);


} );

```

File: src/main/app/controller/courts/LocalAuthoritiesController.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    res.render('courts/tabs/localAuthoritiesContent', pageData);
+  }
+
+
+  public async getLocalAuthorities(
+    req: AuthedRequest,
+    res: Response,
+    updated = false,
+    error = '',
+    selectedAreaOfLaw = '',
```
Comment (amstevenson) [inline src/main/app/controller/courts/LocalAuthoritiesController.ts:75]: I don't think this variable is used in the method below. 

Closest thing to it is: 

```
    const pageData: LocalAuthoritiesPageData = {
      errorMsg: error,
      updated: updated,
      selectedAreaOfLaw: areaOfLaw, < this bit, which is areasOfLaw, not selectedAreaOfLaw
      courtAreasOfLaw: this.getAreasOfLawForSelect(areasOfLaw, areaOfLaw),
      localAuthoritiesItems: this.mapLocalAuthorityToLocalAuthorityItem(allLocalAuthorities, courtLocalAuthorities)
    };
```

File: src/main/app/controller/courts/LocalAuthoritiesController.ts (hunk 2 above)
Comment (Majid-justice) [reply src/main/app/controller/courts/LocalAuthoritiesController.ts]: I have moved enum to its own folder under utils

File: src/main/app/controller/courts/LocalAuthoritiesController.ts (hunk 3 above)
Comment (Majid-justice) [reply src/main/app/controller/courts/LocalAuthoritiesController.ts]: removed all spaces

File: src/main/app/controller/courts/LocalAuthoritiesController.ts (hunk 4 above)
Comment (Majid-justice) [reply src/main/app/controller/courts/LocalAuthoritiesController.ts:71]: refactored code as your suggestion

File: src/main/app/controller/courts/LocalAuthoritiesController.ts (hunk 5 above)
Comment (Majid-justice) [reply src/main/app/controller/courts/LocalAuthoritiesController.ts]: removed variable

File: src/main/assets/js/local-authorities.ts (hunk 1 above)
Comment (Majid-justice) [reply src/main/assets/js/local-authorities.ts]: I have removed the logic to check SuperAdmin from the local-authorities.ts to the edit-court-general.njk using the IsSuperAdmin provided by the middleware

File: src/main/app/controller/courts/LocalAuthoritiesController.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          value: JSON.stringify(la),
+          text: la.name,
+          checked: courtLocalAuthorities.some(e => e.id === la.id)
+        }));
+
+      return localAuthorityItems.sort((a, b) => (a.text < b.text ? -1 : 1));
+    }
+    else
+    {
+      return [];
```
Comment (amstevenson) [inline src/main/app/controller/courts/LocalAuthoritiesController.ts:137]: Is this an error scenario? If so, is it worth adding in a log error here? Not a problem if it's default behaviour 

If it's default though, can always do something like: 

```
    if(courtLocalAuthorities) {
}

> if we find this is an error case, which it doesn't look like it is - log.error('this was not expected') 
return [];
```

File: src/main/assets/js/local-authorities.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+      this.getLocalAuthorities();
+    });
+
+
+  }
+
+
+
+
+
```
Comment (amstevenson) [inline src/main/assets/js/local-authorities.ts:122]: maaaaaaasive white spaces haha

Comment (Majid-justice) [reply src/main/assets/js/local-authorities.ts]: removed lines

File: src/main/app/controller/courts/LocalAuthoritiesController.ts (hunk 6 above)
Comment (Majid-justice) [reply src/main/app/controller/courts/LocalAuthoritiesController.ts:138]: its default behaviour

**Review by amstevenson — APPROVED** [review]

**Pull request merged**